src/common/opt_partition_builder.h

```diff
@@ -440,72 +401,48 @@ class OptPartitionBuilder {
         uint32_t th_end = std::min(th_begin + thread_size, min_node_size);
         if (th_end > th_begin) {
           CHECK_LT(min_node_id, partitions.size());
-          threads_addr[tid].push_back({row_indices_ptr, partitions[min_node_id].b + th_begin,
+          tm.threads[tid].addr.push_back({row_indices_ptr, partitions[min_node_id].b + th_begin,
                                        partitions[min_node_id].b + th_end});
-          threads_id_for_nodes[min_node_id].push_back(tid);
-          node_id_for_threads[tid].push_back(min_node_id);
+          tm.nodes[min_node_id].threads_id.push_back(tid);
+          tm.threads[tid].nodes_id.push_back(min_node_id);
         }
       }
     } else if (n_features*summ_size / n_threads <
                (static_cast<size_t>(1) << (depth + 1))*n_bins
                || (depth >= 1 && !hist_fit_to_l2)) {
       uint32_t block_size = std::max(common::GetBlockSize(summ_size, n_threads),
                                      std::min(summ_size, static_cast<uint32_t>(512)));
-      uint32_t node_id = 0;
       uint32_t curr_thread_size = block_size;
       uint32_t curr_node_disp = 0;
-      uint32_t curr_thread_id = 0;
+      tm.threads_cbuffer.Reset();
+      auto thread_info = tm.threads_cbuffer.GetItem();
       for (uint32_t i = 0; i < n_threads; ++i) {
         while (curr_thread_size != 0) {
-          const uint32_t curr_thread_node_size = threads_nodes_count[curr_thread_id%
-                                                                     n_threads][node_id];
+          uint32_t node_id = tm.threads_cbuffer.NCycles();
```

Comment:
suggest renaming of NCycles with smth like CycleIdx or BlockIdx, it should be clearer.